llvm/labs/lab2/durandin_vladimir/vladimir_inline_pass.cpp

```diff
@@ -0,0 +1,165 @@
+      Instruction *InsertPt = getNextNonTerminator(InsertBB, CI);
+
+      // Create a map for block correspondence
+      DenseMap<BasicBlock *, BasicBlock *> BlockMap;
+
+      // Split the calling function at the call site (use ".splited." for
+      // clarity)
+      BasicBlock *SplitBB = InsertBB->splitBasicBlock(
+          InsertPt, InsertBB->getName() + ".splited." +
+                        std::to_string(counter_splited++));
+
+      // Create new blocks corresponding to empty function's blocks
+      BasicBlock *CalleeEntryBasicBlockPointer = nullptr;
+      for (BasicBlock &CalleeBB : *Callee) {
+        BasicBlock *NewBB = BasicBlock::Create(
+            F.getContext(),
+            CalleeBB.getName() + ".inlined." + std::to_string(counter_inlined),
+            &F);
+        if (Callee->getEntryBlock().getName() == CalleeBB.getName())
+          CalleeEntryBasicBlockPointer = NewBB;
+        BlockMap[&CalleeBB] = NewBB;
+      }
+      InsertBB->getTerminator()->setSuccessor(0, CalleeEntryBasicBlockPointer);
+
+      counter_inlined++;
+
+      // Copy instructions from corresponding blocks
+      ValueToValueMapTy VMap;
+      for (BasicBlock &CalleeBB : *Callee) {
+        BasicBlock *NewBB = BlockMap[&CalleeBB];
+        for (Instruction &Inst : CalleeBB) {
+          IRBuilder<> Builder(
+              NewBB); // Create IRBuilder with specified basic block
+          Instruction *NewInst = Inst.clone(); // Clone the instruction
+          Builder.Insert(NewInst); // Add the cloned instruction to the end of
+                                   // the basic block
+          VMap[&Inst] = NewInst;
+        }
+      }
+
+      // Update branch instructions to target correct blocks
+      // for (Instruction &Branch : *SplitBB) {
+      for (BasicBlock &CalleeBB : F) {
+        for (Instruction &Branch : CalleeBB) {
+          if (auto *BI = dyn_cast<BranchInst>(&Branch)) {
+            for (size_t i = 0, e = BI->getNumSuccessors(); i != e; ++i) {
+              BasicBlock *Successor = BI->getSuccessor(i);
+              if (BlockMap[Successor] != nullptr) {
+                BI->setSuccessor(i, BlockMap[Successor]);
+              }
+            }
+          }
+        }
+      }
+
+      for (BasicBlock &CalleeBB : *Callee) {
+        BasicBlock *NewBB = BlockMap[&CalleeBB];
+        for (Instruction &Inst : *NewBB) {
+          RemapInstruction(&Inst, VMap, RF_IgnoreMissingLocals, nullptr,
+                           nullptr);
+        }
+      }
+
+      // Create IRBuilder and set its context
+      IRBuilder<> Builder(F.getContext());
+
+      // Traverse the basic blocks in Caller
+      for (BasicBlock &BB : F) {
+        // Create a copy of the instruction list for the current basic block
+        std::vector<Instruction *> Instructions;
+        for (Instruction &I : BB) {
+          Instructions.push_back(&I);
+        }
+
+        for (Instruction *I : Instructions) {
+          if (auto *RI = dyn_cast<ReturnInst>(I)) {
```

Comment:
If you need ReturnInst only than maybe save only such instructions into `Instructions` vector. By the way, you may use something like `llvm::SmallVector` container for that.

Reply:
Yes, it makes sense to do that, thank you! 
P.S. I fixed other things too!